app/services/contributors_generator.rb

```diff
@@ -0,0 +1,243 @@
+
+      return [FORM_EVENT]
+    end
+
+    []
+  end
+
+  private
+
+  sig { returns(Work) }
+  attr_reader :work
+
+  sig { params(contributor: Contributor).returns(T.nilable(Cocina::Models::Contributor)) }
+  def contributor(contributor)
+    # FIXME: TODO: mappings for status (primary) and/or order
+    Cocina::Models::Contributor.new(
+      "name": name_descriptive_value(contributor),
+      "type": contributor_type(contributor),
+      "role": cocina_roles(contributor.role)
+    )
+  end
+
+  sig { params(contributor: Contributor).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def name_descriptive_value(contributor)
+    return [Cocina::Models::DescriptiveValue.new(value: full_name(contributor))] unless contributor.person?
+
+    [Cocina::Models::DescriptiveValue.new(value: full_name(contributor), type: 'inverted full name')]
+  end
+
+  sig { params(contributor: Contributor).returns(T.nilable(String)) }
+  def full_name(contributor)
+    return "#{contributor.last_name}, #{contributor.first_name}" if contributor.person?
+
+    contributor.full_name
+  end
+
+  sig { params(contributor: Contributor).returns(String) }
+  def contributor_type(contributor)
+    return 'conference' if contributor.role == 'Conference'
+    return 'event' if contributor.role == 'Event'
+
+    contributor.contributor_type
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  sig { params(role: String).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def cocina_roles(role)
+    result = [
+      Cocina::Models::DescriptiveValue.new(
+        value: role,
+        source: {
+          value: 'Stanford self-deposit contributor types'
+        }
+      )
+    ]
+    result << T.must(marcrelator_role(role)) if marcrelator_role(role)
+    result << T.must(datacite_role(role)) if datacite_role(role)
+    result
+  end
+  # rubocop:enable Metrics/MethodLength
+
+  # rubocop:disable Metrics/MethodLength
+  sig { params(role: String).returns(T.nilable(Cocina::Models::DescriptiveValue)) }
+  def marcrelator_role(role)
+    mr_code = ROLE_TO_MARC_RELATOR_CODE[role]
+    mr_value = MARC_RELATOR_CODE_TO_VALUE[mr_code]
+    return unless mr_code || mr_value
+
+    Cocina::Models::DescriptiveValue.new(
+      value: mr_value,
+      code: mr_code,
+      uri: "http://id.loc.gov/vocabulary/relators/#{mr_code}",
+      source: {
+        code: 'marcrelator',
+        uri: 'http://id.loc.gov/vocabulary/relators/'
+      }
+    )
+  end
+  # rubocop:enable Metrics/MethodLength
+
+  # rubocop:disable Metrics/MethodLength
+  sig { params(role: String).returns(T.nilable(Cocina::Models::DescriptiveValue)) }
+  def datacite_role(role)
+    datacite_role = ROLE_TO_DATA_CITE_VALUE[role]
+    return if datacite_role.blank?
+
+    case datacite_role
+    when String
+      datacite_source = 'DataCite contributor types'
+    when :creator
+      datacite_role = datacite_role.to_s.capitalize
+      datacite_source = 'DataCite properties'
+    end
```

Comment:
This could be extracted to another method to make this one shorter, since it seems to have two responsibilities: looking up a role string to derive role and source values, and then injecting those values into a new DV. But I don't think this method is hard to grok, so I'm fine with it staying as it with the inline exclusions. :+1:

Reply:
yeah, i thought about it but deriving two values in a single method starts to need a data structure and that gets harder to grok.